### PR TITLE
AB#152791: Enforces one structure one identity within a workbook

### PR DESCRIPTION
**Pull request by JoeDavies-6**
# Overview

- No duplicate structures within workbook or workbook and compound table
- Additional logic for adding novel compounds
- Refactored novel compound code into services
- Reintroduced STOUT for iupac naming and updated toml and lock

## Azure Boards

- [AB#152791](https://dev.azure.com/UniversityOfNottingham/bd2de5ba-2a41-4bdc-a444-61e6b706132a/_workitems/edit/152791)

### CODE REVIEW

**Comment by JoeDavies-6** [comment]
Hey @AndyRae, we unsurprisingly fail on 3.10 with the updates to the .toml. Is there anything you think should anything be done?

File: poetry.lock
Comment (JoeHeeley2) [inline poetry.lock:1]: It looks like a few packages have been updated here, is this expected behaviour of poetry? I'm slightly worried something might break if there's a change in dependency we haven't accounted for

Comment (JoeDavies-6) [reply poetry.lock:1]: I've found one of these - werkzeug - I think flask incorrectly labels the required version which may not help. I'll add a new commit to fix this. https://stackoverflow.com/questions/77213053/why-did-flask-start-failing-with-importerror-cannot-import-name-url-quote-fr

**Comment by AndyRae** [comment]
> Hey @AndyRae, we unsurprisingly fail on 3.10 with the updates to the .toml. Is there anything you think should anything be done?


Looks like there's been some commits since you asked, but the dependencies currently fail:
`The currently activated Python version 3.10.13 is not supported by the project (>3.9,<3.10)` 

It worked before as: 
`python = "^3.9"` this means any version from 3.9, up to but excluding 4.

But the change:
`python = ">3.9,<3.10"` means versions greater than 3.9 but strictly less than 3.10. 

So the fix is to move back to `^3.9`, or if you can't support it anymore (or don't want to), move up to `^3.10` I built in the matrix tests to support moving up easier

**Comment by JoeDavies-6** [comment]
> > Hey @AndyRae, we unsurprisingly fail on 3.10 with the updates to the .toml. Is there anything you think should anything be done?
> 
> Looks like there's been some commits since you asked, but the dependencies currently fail: `The currently activated Python version 3.10.13 is not supported by the project (>3.9,<3.10)`
> 
> It worked before as: `python = "^3.9"` this means any version from 3.9, up to but excluding 4.
> 
> But the change: `python = ">3.9,<3.10"` means versions greater than 3.9 but strictly less than 3.10.
> 
> So the fix is to move back to `^3.9`, or if you can't support it anymore (or don't want to), move up to `^3.10` I built in the matrix tests to support moving up easier

Hey Andy, thanks that makes more sense now. I've made a change now which I think restores us to the previous situation of supporting 3.9 and 3.10. If the change looks good to you we can go ahead with the PR. If not, I'll look at moving up to 3.10 instead

**Review by AndyRae — APPROVED** [review]

**Review by JoeHeeley2 — APPROVED** [review]

**Pull request merged**